first push - bwf specification for badminton court

AimanFariz opened the pull request:
Hi Ross,

This is my work to add Badminton to this package. Related Open Issue is https://github.com/sportsdataverse/sportyR/issues/41

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added support for plotting badminton courts, including official Badminton World Federation (BWF) dimensions and customizable court features.
  - Introduced functions to generate geometric representations of badminton court areas, lines, and net for visualization.
  - Added a function to customize colors for various badminton court features.

- **Documentation**
  - Updated README files to include badminton as a supported sport, with new details on plotting options and general manager information.

- **Tests**
  - Added tests to ensure badminton court plots are generated correctly.

- **Chores**
  - Updated internal data to include official and customizable badminton court dimensions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

AimanFariz committed:
first push - bwf specification for badminton court

rossdrucker reviewed (changes requested):
Thanks for the work here so far @AimanFariz. I've reviewed the PR, and still feel there's more to be done before this can be incorporated into `{sportyR}`.

Tl;dr – while an unrotated plot does seem to be generated correctly, I have some larger concerns with some of the underlying mechanics/parameterizations. I don't feel this code is ready to merge or slate for release yet. I'm **requesting changes** as a result

To summarize the general themes:

1. **Surface dimensions** – the rest of the package uses full-surface dimensions in the parameterization of each playing surface. It seems that for `bwf`, there is a mix. I called it out in a comment, but to me this is the primary thing to address first: in `surface-dimensions.json`, the `court_length` and `court_width` consistently refer to the full-sized dimensions. Once this is updated, I'm assuming that you'll need to refactor some of the other code you're introducing here to keep the generated plot in line
2. **Parameterization consistency** – within the `"custom"` parameters, please ensure that you're capturing _all_ surface parameters that are intended to be used by the functions
3. **Output of** `geom_badminton()` – this should _only_ output the plot, and none of the underlying surface dimensions should be printed/displayed
4. **Raw feature definition** – in general, please try to align the edge of the raw features with where their edges lie on the court. As an example, the endline of a basketball court is entirely out of play, so its inner edge is anchored along `x = 0` and then is moved to the edge of the court when adding to the plot. Please double-check that you've followed this pattern for all features
5. **Internal consistency** – I noticed several places where your code is not internally consistent (e.g. "frontcourt" vs. "forecourt"). Please double-check these inconsistencies
6. **Testing** – as noted in the third bullet of my response to you [here](https://github.com/sportsdataverse/sportyR/issues/41#issuecomment-2888834412), a `geom_badminton()` test case needs to be added [here](https://github.com/sportsdataverse/sportyR/blob/main/tests/testthat/test-no-league-error.R) to ensure code coverage. Following other sports' test code should be a good guide, and the code that checks this is provided [here](https://github.com/sportsdataverse/sportyR/blob/main/tests/testthat/test-no-league-error.R). One other helpful item would be to add a test into `tests/testthat/test-cani-functions.R` to test if a single-league sport message is correctly printed. I can help with this if needed
7. **Rotational variance** – when manually testing the code you're introducing here to see if the plots change when rotating, there is a variance of your plot. While it may be unlikely that a user supplies an angle of 45º to a plot, the generated plot should nevertheless be identical to if a user were to tilt their head. This is not upheld (see the example I provided in my comments)
8. **Code styling** – there are a few things that stood out to me here. I don't mean to nitpick these points, but the internally-consistent code styling really helps to expand the package quickly and follow maintenance items through. A few things to double-check:

- Line lengths should not exceed 80 characters
- There should be 5 blank lines at the end of each section
- Spaces around infix operators
- Section and subsection headers should match throughout the code you're introducing, and should be human-readable (e.g. no `_` in these headings)
- Comments should go above the line that they refer to, not after
- Prefer `"` to `'`
- Specify colors via hexidecimal values (e.g. prefer `"#ffffff"` to `"white"`)

AimanFariz commented on sportyR.Rproj:
```diff
@@ -1,4 +1,5 @@
 Version: 1.0
+ProjectId: 588338e4-1e15-4f3e-9b9c-c7e6ec7fd161
```
Sure, I'm not sure how it got there. I did not manually add it there.

AimanFariz commented on data-raw/surface-dimensions.json:
```diff
@@ -1,4 +1,33 @@
+
+        "bwf": {
+            "court_length": 13.4,
+            "singles_width": 2.53,
```
yes, you're right. This is definitely another oversight and I will change this

AimanFariz commented on R/geom-badminton.R:
```diff
@@ -0,0 +1,559 @@
+    reflect_y = FALSE,
+    x_trans = x_trans,
+    y_trans = y_trans,
+    rotation = -rotation
```
This is a miscalculation from my side. I will reflect the correct change in a later commit.

AimanFariz commented on R/features-badminton.R:
```diff
@@ -0,0 +1,251 @@
+#'   forecourt
+#'
+#' @keywords internal
+badminton_frontcourt_half <- function(short_serviceline_distance = 0,
```
It is supposed to be forecourt. I initially thought it's called frontcourt, as in tennis. Further checks reveal that it's forecourt. I will make sure to change this and ensure consistency.

AimanFariz commented on R/features-badminton.R:
```diff
@@ -0,0 +1,251 @@
+
+#' The back boundary line is the line behind which a player will serve the ball. It spans
+#' the entire width of the court, and its back edge denotes the furthest
+#' boundary inside of which a ball can land and be considered in play
```
Yes, it's either "Shuttlecock" or "Birdie". This is one of the things that I forgot to change from copying the geom-tennis.R file.

I will change all the 'ball' instances to 'shuttlecock' in a later commit. Thanks for pointing this out.

AimanFariz commented on R/geom-badminton.R:
```diff
@@ -0,0 +1,559 @@
+
+  # Set Display Range-----------------------------------------------------------
+  half_court_length <- ((court_params$court_length %or% 0) / 2) +
+    (court_params$backstop_distance %or% 20) +
```
No, It should just be 2m, and the sidesrop_distance should be just 1m instead of 10. Thanks for pointing this out

AimanFariz committed:
2nd push - fixed errors and made changes from first push

AimanFariz commented:
The 2nd commit is meant to reflect all the changes. Let me know if there's anymore issues with the files.

rossdrucker commented on R/features-badminton.R:
```diff
@@ -1,38 +1,36 @@
-#'
-#' This is one half of the forecourt (either the left or right service court)
+#' Badminton Forecourt Half
+#' The forecourt is the area between the [badminton_net()] and the
```
Can you add a new line here (and for all other functions below)? Example:

```r
#' Badminton Forecourt Half
#' 
#' The forecourt is the area between [...]
```

rossdrucker commented on R/features-badminton.R:
```diff
@@ -1,38 +1,36 @@
 #' @keywords internal
-badminton_frontcourt_half <- function(short_serviceline_distance = 0,
+badminton_forecourt_half <- function(short_serviceline_distance = 0,
                                    singles_width = 0) {
```
Please keep the argument names aligned. See other files in the package for examples as needed

AimanFariz replied:
commenting to keep track of changes. Done

rossdrucker commented on R/features-badminton.R:
```diff
@@ -1,38 +1,36 @@
 #' @param court_length The length of the court
-#' @param long_serviceline_distance The distance from the net to the short service line
+#' @param long_serviceline_distance The distance from the net to the short
+#' service line
```
I believe this should be indented as its a continuation of the parameter definition

AimanFariz replied:
Done

rossdrucker commented on R/features-badminton.R:
```diff
@@ -122,7 +118,9 @@ badminton_sideline <- function(court_length = 0, feature_thickness = 0) {
 
-badminton_centerline <- function(court_length = 0, short_serviceline_distance = 0, feature_thickness = 0) {
+badminton_centerline <-
+  function(court_length = 0, short_serviceline_distance = 0,
```
Please move the function definition back up to be in line with the `<-`, and then move each argument after `court_length` to its own new line. Commenting here, but applies to the rest of the PR

rossdrucker commented on R/geom-badminton.R:
```diff
@@ -418,14 +416,12 @@ geom_badminton <- function(league,
   )
 
-  # Set Display Range-----------------------------------------------------------
+  # Set Display Range -----------------------------------------------------------
```
I believe you're one character over here; delete one of the `-` and it should be fine/consistent

AimanFariz replied:
Done

rossdrucker commented on R/geom-badminton.R:
```diff
@@ -0,0 +1,559 @@
+#'
+#' @param plot_background A hexadecimal string representing the color to use for
+#'   this feature
+#' @param back_boundary_line A hexadecimal string representing the color to use for this
```
In RStudio, for commented lines, you an highlight the commented line and go to `Code > Reflow Comment`

rossdrucker reviewed (changes requested):
@AimanFariz there still seems to be some out-standing comments that I'd like to have addressed before merging this PR. Please confirm that your code style matches the rest of the package internally as well.

Also, is there a reason that there are graphics updates here?

If it's helpful, feel free to make more than one commit to address these items rather than trying to accomplish it all in one commit

AimanFariz commented:
I'll get the comments resolved in a later commit. 

Regarding the graphics update, I believe it came from the README.Rmd. I was just changing the README to make sure it addresses the spacing issue from the first push. I saved the file with "Knit on Save", which I thought must be done to ensure the change is reflected in the README.md too. I might be wrong there. What do I have to do with the graphics now? Do I just leave them be?

AimanFariz committed:
push three - fix changes from push 2

rossdrucker commented:
@AimanFariz apologies for the delay in re-reviewing this PR. I just made some additional comments that hopefully help get towards merging. If you don't mind as well, can you please address them in smaller commits? It's easier for me to review and track changes that way instead of as one big commit. Let me know if anything's unclear, and thanks for the continued work here!

AimanFariz committed:
Merge branch 'sportsdataverse:main' into badminton-aiman

AimanFariz commented:
I'm still working on the changes. The merge request should not be approved yet. I was trying to "Sync Fork" the recent changes from the main SportyR to the branch that I'm working on.

AimanFariz committed:
commit 4 - fix changes from 3, details in the Comments

AimanFariz commented:
Summary from Commit 4:
- Badminton Forecourt title change
- backcourt calculation from features-badminton
- Added "Badminton Feature Colors" title
- Added "Draw Badminton Court" title
- Fixed indentation in geom_badminton()
- matched ylim and xlim calculation in geom_badminton to only be "serve" and "receive"
- Added surface line definition for long service lines

Another comment(s):
- I tried deleting extra graphics from a previous commit, and although git status shows that the graphics were deleted, running git add . did not stage the deletion. The same issue happened when adding the geom-badminton test case in test-no-league-error.R, where the new test shows up in git status but does not get staged when I run git add ..
